shiny/ui/_theme.py

```diff
@@ -356,6 +375,81 @@ def add_rules(
         self._reset_css()
         return self
 
+    def add_sass_layer_file(self: T, path: str | pathlib.Path) -> T:
+        """
+        Add a Sass layer file to the theme.
+
+        This method reads a special `.scss` file formatted with layer boundary comments
+        to denote regions of functions, defaults, mixins, and rules. It then splits the
+        file into these constituent pieces and adds them to the appropriate layers of
+        the theme.
+
+        The theme file should contain at least one of the following boundary comments:
+
+        ```scss
+        /*-- scss:uses --*/
+        /*-- scss:functions --*/
+        /*-- scss:defaults --*/
+        /*-- scss:mixins --*/
+        /*-- scss:rules --*/
+        ```
+
+        Each layer, once extracted, is added to the theme using the corresponding
+        `add_` method, e.g. the `scss:rules` layer is added via `.add_rules()`.
+
+        Layer types can appear more than once in the `.scss` file. They are coalesced
+        into a single layer by order of appearance and then added as a block via their
+        corresponding `add_` method.
+
+        Parameters
+        ----------
+        path
+            The path to the `.scss` file to be added.
+
+        Raises
+        ------
+        ValueError
+            If the `.scss` file doesn't contain at least one valid region decorator.
+        """
+        with open(path, "r") as file:
+            src = file.readlines()
+
+        layer_keys = ["uses", "functions", "defaults", "mixins", "rules"]
+        rx_pattern = re.compile(rf"^/\*--\s*scss:({'|'.join(layer_keys)})\s*--\*/$")
+
+        layer_boundaries = [rx_pattern.match(line.strip()) for line in src]
+
+        if not any(layer_boundaries):
+            raise ValueError(
+                f"The file {path} doesn't contain at least one layer boundary "
+                f"(/*-- scss:{{{','.join(layer_keys)}}} --*/)",
+            )
+
+        layers: dict[str, list[str]] = {}
+        layer_name: str = ""
+        for i, line in enumerate(src):
+            layer_boundary = layer_boundaries[i]
+            if layer_boundary:
+                layer_name = layer_boundary.group(1)
+                continue
+
+            if not layer_name:
+                # Preamble lines are dropped (both in Quarto and {sass})
+                continue
+
+            if layer_name not in layers:
+                layers[layer_name] = []
+
+            layers[layer_name].append(line)
+
+        for key, value in layers.items():
+            # Call the appropriate add_*() method to add each layer
+            add_method = getattr(self, f"add_{key}", None)
+            if add_method:
+                add_method("".join(value))
```

Comment:
I don't feel strongly about it (especially since there are tests), but I'd lean towards throwing an error if the method doesn't exist